pkg/pubsubsource/pubsub_test.go

```diff
@@ -0,0 +1,221 @@
+package pubsubsource
+
+/*
+Copyright 2022 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/numaproj/numaflow-go/pkg/sourcer"
+	"github.com/ory/dockertest/v3"
+	"github.com/ory/dockertest/v3/docker"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/numaproj-contrib/gcp-pub-sub-source-go/pkg/mocks"
+)
+
+const TopicID = "pubsub-test"
+const Project = "pubsub-local-test"
+const subscriptionID = "subscription-09098ui1"
+
+var (
+	pubsubClient *pubsub.Client
+	resource     *dockertest.Resource
+	pool         *dockertest.Pool
+)
+
+func publish(ctx context.Context, client *pubsub.Client, topicID, msg string) error {
+	t := client.Topic(topicID)
+	result := t.Publish(ctx, &pubsub.Message{Data: []byte(msg)})
+	_, err := result.Get(ctx)
+	if err != nil {
+		return fmt.Errorf("pubsub: result.Get: %w", err)
+	}
+	return nil
+}
+
+func sendMessages(ctx context.Context, client *pubsub.Client, topicID string, count int) {
+	for i := 0; i < count; i++ {
+		if err := publish(ctx, client, topicID, fmt.Sprintf("Message %d", i)); err != nil {
+			log.Fatalf("Failed to publish: %v", err)
+		}
+	}
+}
+
+func ensureTopicAndSubscription(ctx context.Context, client *pubsub.Client, topicID, subID string) error {
+	topic := client.Topic(topicID)
+	exists, err := topic.Exists(ctx)
+	if err != nil {
+		return err
+
+	}
+	if !exists {
+		if _, err = client.CreateTopic(ctx, topicID); err != nil {
+			return err
+		}
+	}
+
+	sub := client.Subscription(subID)
+	exists, err = sub.Exists(ctx)
+	if err != nil {
+		return err
+
+	}
+	if !exists {
+		if _, err = client.CreateSubscription(ctx, subID, pubsub.SubscriptionConfig{Topic: topic}); err != nil {
+			return err
+
+		}
+	}
+	return nil
+}
+
+func TestMain(m *testing.M) {
```

Comment:
This unit test includes multiple go-rountines. Please use https://github.com/uber-go/goleak to detect any potential leakages. Just add `goleak.VerifyTestMain(m)` to your TestMain func.

Reply:
there are third party api go routine leakages ,which is difficult to fix. For the custom go routines have added the context with timeout and cancel

Reply:
We shouldn't introduce dependencies that have goroutine leakages, such bad dependencies will come to bite us at some point. Which third-party api are we talking about? is there any alternatives?

Reply:
I found this issue related to the internal library which is indirect dependency that is causing the go routine leak -https://github.com/census-instrumentation/opencensus-go/issues/1191

Reply:
https://github.com/numaproj-contrib/gcp-pub-sub-source-go/issues/3